### DISCUSSION TITLE
HTTP Request for BLE Sensors (Temperature)

### DISCUSSION THREAD

**Discussion by Celpear** [arendst/Tasmota#15608] [Q&A] [answered]
Hello,
I've been trying to parse the data displayed on the main page of Tashmota for a while now. I get the data via /?m=1 in a strange format. The API description does not help me. And also the Google search was not helpful. Is there a console command to get the data?
![Unbenannt](https://user-images.githubusercontent.com/28110865/168159217-8924628c-64f8-4673-bba3-24beed2746af.png)

![data](https://user-images.githubusercontent.com/28110865/168159487-98badd6a-0982-4037-8c69-9f79034ffe00.png)

Thanks for the help!

**Comment by sfromis** [comment, marked as answer]
> I get the data via /?m=1

That is no API and not a supported way of doing it. The sensor values should be included in the JSON-formatted output of the command `Status 10` available via the regular documented http command interface.

**Reply by Celpear** [reply]
Thanks for your answer!
Status 10 I had also already tried. Unfortunately, I do not get the BLE temperature data.

07:44:44.794 CMD: Status 10
07:44:44.808 RSL: STATUS10 = {"StatusSNS":{"Time": "2022-05-13T07:44:44", "ESP32":{"Temperature":45.0}, "TempUnit": "C"}}

It does show a heartbeat message in the console, with the desired data. But I don't know how to trigger it.
<img width="1376" alt="Bildschirmfoto 2022-05-13 um 08 46 14" src="https://user-images.githubusercontent.com/28110865/168228312-145f9d59-69bb-47d7-9721-aa630a962e28.png">



My HTTP GET request:
![request](https://user-images.githubusercontent.com/28110865/168228052-88803d04-0145-47d4-9e36-901582bf2526.png)

Thanks for all your help!

**Reply by sfromis** [reply]
Inclusion of BLE sensor values in the `...SENSOR` payloads and the output of `Status 10` may depend on your selected [MI32Option](https://tasmota.github.io/docs/Commands/#mi32option) values. I think I have used `MI32Option2 1`.

I positively verified having the values appearing before posting.
`{"StatusSNS":{"Time":"2022-05-13T10:02:16.908+02:00","LYWSD03-xxxxxx":{"Temperature":26.30,"Humidity":40.70,"DewPoint":11.86,"RSSI":-58},"TempUnit":"C"}}`
(I killed off the useless ESP32 Temperature reading using `SetSensor127 0`)

**Reply by HawkaiV** [reply]
I have the same issue as Celpear, no sensor temperature in STATUS 10. SENSOR show correct values. I'm using a LYWSDxxx Xiaomi bluetooth sensor. I've followed sfromis instructions but no luck. 

Using latest bluetooth release on an ESP32

Any more suggestions that could help?

**Reply by barbudor** [reply]
As pointed by sfromis, have you looked at mi32option ?
What are your current mi32option settings ?

**Reply by barbudor** [reply]
Ah correct, valid in SENSOR but missing in Status 10

**Reply by barbudor** [reply]
I can confirm that the BLE driver cannot support Status 10
The function to append to the SENSOR / Status 10 JSON is disabled and this since the inception of the driver
https://github.com/arendst/Tasmota/blob/5d3f2cc316892f7034407bdf7c23a99077c7d102/tasmota/tasmota_xsns_sensor/xsns_62_esp32_mi_ble.ino#L3553-L3556

From the beginning, this driver has always published it's SENSOR message on it's own way, not following Tasmota standards
One of the reason is that as a gateway, the number of sensors could be very large and override the capacity of the buffers.

I have no idea how @sfromis could have add that result but that can't be with this driver

**Reply by HawkaiV** [reply]
edit: Ah, sorry missed you replies while writing my answer. Problem "solved"

Thank you for your answer @barbudor 
Yes I did. Mi32option2 1 don't solve it. Also tried Mi32option2 0 with no difference in output
I've tried Mi32option(x) with both 0 and 1 (x=0...4) 

![capture](https://user-images.githubusercontent.com/98651179/217952246-79d47970-790d-46db-908b-e7605c246e05.png)

**Reply by barbudor** [reply]
@sfromis you seem to be using the other driver `xsns_62_esp32_mi.ino` and not `xsns_62_esp32_mi_ble.ino`

Are you using the homekit build ?

**Reply by barbudor** [reply]
@HawkaiV I don't understand why you say "Problem solved" but confirm that Mi32Option doesn't change anything

You may want to try the homekit build which seems to use the other driver
Those 2 drivers are confusing...

**Reply by sfromis** [reply]
Not sure what is happening..... I tested again with `Status 10`, and no longer get the LYWSD03 values appearing.

**Reply by HawkaiV** [reply]
@barbudor, Correct, I meant that we have a reason for why it's not working, but it's not solved of course

I'll try the homekit build. I just need to figure it out how to do it. I'll be back :-)

**Reply by barbudor** [reply]
There are 2 drivers
 - xsns_62_esp32_mi_ble.ino is used in tasmota32-bluetooth and is the btsimon's one => do not support FUNC_JSON_APPEND
 - xsns_62_esp32_mi.ino is used in tasmota32-homekit (is that staar's) => do supoprt FUNC_JSON_APPEND

So confusing

**Reply by sfromis** [reply]
> Are you using the homekit build ?

Could be at that time, not really sure now. Not been focusing on BLE, more or less given up on having so much trouble.

INFO1 says `"Version":"11.1.0.2(bluetooth)` but that does not tell that much.

**Reply by barbudor** [reply]
@HawkaiV You can flash the homekit build from the web installer, or grab the binary from : https://github.com/Jason2866/Tasmota-specials/tree/firmware/firmware/tasmota32/other

**Reply by HawkaiV** [reply]
@barbudor When you say the homekit build, which do you mean? No build is called homekit. Homebridge is the closest one

**Reply by barbudor** [reply]
Sorry, yes `tasmota32-mi32-homebridge` is the name of the build

![image](https://user-images.githubusercontent.com/2996042/217954884-f0230dde-033e-46c5-84ee-8e86f368d24c.png)

**Reply by HawkaiV** [reply]
YIPPIE!! Many thanks for your help @barbudor! I'll try this one for a while and see if there's any issues with the build
![Skärmbild 2023-02-09 234326](https://user-images.githubusercontent.com/98651179/217956120-79573a43-6b45-4252-8156-ba013cda814d.png)

The purpose of this is to get the value into Samsung Smartthings. The edge driver in Smartthings needs the name and the temperature of the sensor, now it's probably possible! :-)

**Reply by sfromis** [reply]
Tried with the Homebridge build now, and since command `BLEAlias` is unknown, I don't think that I was trying it at that time.

**Reply by barbudor** [reply]
There's doesn't seem to be an alias in your posted StatusSNS message
The Github's Blame show that the commented call I have highlighted has been there for 2 years, so likely from the first commit.

**Comment by AloisKlingler** [comment]
sorry for answering here to an solved request - but it does not work for me.

I have installed tasmota32 MI BLE (the mi homebridge is not available any more).

I can see the sensor correctly:
![image](https://github.com/arendst/Tasmota/assets/11095760/3cdcc00d-9efe-4460-9a9e-6da58bfb08b9)

I it does not show on status 10 the sensor values:
`{"StatusSNS":{"Time":"2023-11-24T18:52:47","MJ_HT_V1-35e58a":{"Battery":89,"RSSI":-97,"MAC":"582D3435E58A"}}}`

why?

Thanks!
BR Alois

**Reply by sfromis** [reply]
Because Tasmota cannot "read" BLE sensor values, only report it when the sensor decides to send. This is in contrast to wired sensors, where `Status 10` (and so on) can trigger a cycle to read current sensor values.

**Reply by AloisKlingler** [reply]
Is there a way to fetch the last cached data, as shown on the mi dashboard, as json?

**Reply by barbudor** [reply]
Officially not. This driver does it's own MQTT publishing and there is no official way to retreive the sensor data whether by HTPP or even through MQTT.

Unofficially you can poll the HTML data that is used by the web page by getting `http://ip.of.tasmota.dev/?m=1`
Up to you to parse that
There is no garantuee that this will work this way in the future

You may have to increase you MiPage to insure you have all the sensors in a single page

**Reply by sfromis** [reply]
What I'm doing is to use `MI32Option2 1` to have LYWSD03 data included in the TelePeriod sensor messages published via MQTT. If I then wanted to ensure quick availability of this from a newly subscribing other client, like a home automation backend, without waiting for next telemetry period, I could set `SensorRetain` to have theses messages retained in the broker. No unsupported scraping.

**Reply by AloisKlingler** [reply]
Hm, what a putty. So parsing html is no issue for me, but not nice.
This sounds like nobody created an enhancement request yet?
As it is in unofficial folder - is there a specific process or person to ask, does somebody know?

Thank you both for replying. 👍🙏

**Reply by barbudor** [reply]
You can rename your discussion "Feature Request"
It doesn't mean that someone will do it
You are the one who want to do HTTP polling on a software primarly designed for MQTT
Feel free to propose a PR

**Reply by AloisKlingler** [reply]
if there would be the sensor data in the mqtt data, I would not search it beeing available by HTTP.
![image](https://github.com/arendst/Tasmota/assets/11095760/808ffb9f-8d04-4438-83d8-8a19a5ffaf63)

https://tasmota.github.io/docs/Bluetooth/#features

in this example there is data available. in mine not. :-(

**Reply by barbudor** [reply]
I thought you didn't want to use MQTT

have 20 LYWSD03MMC sensors operating through MQTT across 2 sites and that is working fine.
The only thing is that the driver publish on his own and not through the official SENSOR/StatusSNS mechanism

There's a bunch of Mi32Option to configure to get what you want
My setup is using:

```
00:00:00.214 MI32Option0: 1
00:00:00.215 MI32Option1: 0
00:00:00.215 MI32Option2: 0
00:00:00.215 MI32Option4: 1
00:00:00.216 MI32Option5: 1
00:00:00.226 MI32Option6: 2
```

With BleAlias for each sensor and I get each sensor on a dedicated topic like that:

```
10:26:58.890 MQT: tele/blegateway/chbre_lucas = {"Time":"2023-11-25T10:26:58+01:00","alias":"chbre_lucas","mac":"a4c1382bdb8e","Temperature":18.3,"Humidity":49,"DewPoint":7.2,"Btn":0,"Battery":97,"RSSI":-96,"TempUnit":"C"} (retained)
10:26:59.893 MQT: tele/blegateway/salle_bain = {"Time":"2023-11-25T10:26:59+01:00","alias":"salle_bain","mac":"a4c13860b9f8","Temperature":19.3,"Humidity":53,"DewPoint":9.5,"Btn":1,"Battery":70,"RSSI":-87,"TempUnit":"C"} (retained)
10:27:02.896 MQT: tele/blegateway/pte_piece = {"Time":"2023-11-25T10:27:02+01:00","alias":"pte_piece","mac":"a4c1384c3eff","Temperature":14.8,"Humidity":55,"DewPoint":5.8,"Btn":1,"Battery":51,"RSSI":-84,"TempUnit":"C"} (retained)
10:27:13.910 MQT: tele/blegateway/salon = {"Time":"2023-11-25T10:27:13+01:00","alias":"salon","mac":"a4c138352c48","Temperature":15.4,"Humidity":54,"DewPoint":6.1,"Btn":1,"Battery":70,"RSSI":-89,"TempUnit":"C"} (retained)
```

MQTT works fine in self publish mode (no polling)
You need to find which set of Mi32Option works best for you 
If you use Sensor Retain, previous published value is kept until next is publish

**Reply by AloisKlingler** [reply]
🙄 have flashed now again the original tasmota32-bluetooth and now I get the sensordata by MQTT 👍 

thanks!

![image](https://github.com/arendst/Tasmota/assets/11095760/4a284c66-aa18-4633-9577-9c28e68d9608)
